### PR TITLE
vm: support "vm secret add/delete/show"

### PR DESCRIPTION
**Pull request by yugangw-msft**
```bash
az vm secret add -g myGroup -n myVM --keyvault myKVInTheSameGroup --cert mycert
az vm secret delete -g myGroup -n myVM --keyvault myKVInTheSameGroup --cert mycert
az vm secret show g myGroup -n myVM

```

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py [hunk 1]
```diff
@@ -130,7 +130,11 @@ def transform_sku_for_table_output(skus):
 cli_command(__name__, 'vm resize', custom_path.format('resize_vm'), no_wait_param='no_wait')
 cli_command(__name__, 'vm capture', custom_path.format('capture_vm'))
 cli_command(__name__, 'vm open-port', custom_path.format('vm_open_port'))
-cli_command(__name__, 'vm format-secret', custom_path.format('get_vm_format_secret'))
+cli_command(__name__, 'vm format-secret', custom_path.format('get_vm_format_secret'), deprecate_info='az vm secret format')
+cli_command(__name__, 'vm secret format', custom_path.format('get_vm_format_secret'))
+cli_command(__name__, 'vm secret add', custom_path.format('add_vm_secret'))
+cli_command(__name__, 'vm secret show', custom_path.format('show_vm_secret'))
+cli_command(__name__, 'vm secret delete', custom_path.format('remove_vm_secret'))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py]: Typically in the CLI, create/delete are paired for resources and add/remove are paired for collections.  I would either change the add to create or delete to remove.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py [hunk 2]
```diff
@@ -2116,6 +2116,69 @@ def get_vm_format_secret(secrets, certificate_store=None):
...
+    else:
+        vault_secret_group = VaultSecretGroup(source_vault=SourceVault(keyvault), vault_certificates=[vault_cert])
+        vm.os_profile.secrets.append(vault_secret_group)
+    vm = set_vm(vm)
+    return vm.os_profile.secrets
+
+
+def show_vm_secret(resource_group_name, vm_name):
+    vm = get_vm(resource_group_name, vm_name)
+    return vm.os_profile.secrets
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: If secrets is a collection, then should this command be `list` and not `show`?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py [hunk 3]
```diff
@@ -2116,6 +2116,69 @@ def get_vm_format_secret(secrets, certificate_store=None):
...
+
+
+def remove_vm_secret(resource_group_name, vm_name, keyvault, certificate=None):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: Based on this I would recommend changing `vm secret delete` to `vm secret remove` 🕺

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py [hunk 4]
```diff
@@ -2385,6 +2386,31 @@ def test_create_vm_from_existing_nic(self, resource_group):
         self.assertTrue(re.match(r'[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+', result['publicIps']))
         self.assertTrue(re.match(r'[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+', result['privateIps']))
 
+
+class VMSecretTest(ScenarioTest):
+    @ResourceGroupPreparer()
+    def test_vm_secret_e2e_test(self, resource_group):
+        # a few pre-made key vault artifacts, with the assumption that this test must under team test subscription of 'AzureSDKADGraph2'
+        keyvault = '/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/sdk-test/providers/Microsoft.KeyVault/vaults/vm-secret-kv'
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: Why do we need a premade artifact for this?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py [hunk 5]
```diff
@@ -2385,6 +2386,31 @@ def test_create_vm_from_existing_nic(self, resource_group):
...
+        certificate = 'vm-secrt-cert'
+
+        # under playback, keyvault id need to adjust to use mock subscription
+        test_keyvault_id = re.sub('[0-9A-Fa-f]{8}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{12}', self.get_subscription_id(), keyvault)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: ScenarioTest has a `get_subscription_id` helper method that will automatically return the mocked subscription ID.  You could build your IDs using that and eliminate this regex.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py [hunk 6]
```diff
@@ -2385,6 +2386,31 @@ def test_create_vm_from_existing_nic(self, resource_group):
...
+
+        vm = 'vm1'
+        self.cmd('vm create -g {} -n {} --image rhel --use-unmanaged-disk'.format(resource_group, vm))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: If you create the KeyVault and then the VM, I wouldn't think you'd run into DNS issues to issue further KeyVault commands once the VM finished.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py (hunk 4 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: It is about the keyvault DNS issue you mentioned below. On a second thought, I will just abandon the premade, and create the vault on the fly. Extra 60 seconds looks like is needed even i put it before the vm .

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_help.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
             az {1} --ids $(az vm list -g MyResourceGroup --query "[].id" -o tsv)
 """
 
-helps['vm format-secret'] = """
+helps['vm secret format'] = """
```
[the remark this answers is not in the document]
Comment (tjprescott) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_help.py]: Yes.

[the remark this answers is not in the document]
Comment (tjprescott) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_help.py]: Actually, the history says it is copied, so no, not yet.

**Pull request merged**